tests/jetstream_test.ts

```diff
@@ -3010,3 +3010,97 @@ Deno.test("jetstream - pull next", async () => {
 
   await cleanup(ns, nc);
 });
+
+Deno.test("jetstream - pull errors", async () => {
+  const { ns, nc } = await setup(jetstreamServerConf({}, true));
+
+  const { stream, subj } = await initStream(nc);
+  const jsm = await nc.jetstreamManager();
+
+  await jsm.consumers.add(stream, {
+    durable_name: "me",
+    ack_policy: AckPolicy.Explicit,
+  });
+  const js = nc.jetstream();
+
+  async function expectError(
+    expires: number,
+    code: ErrorCode,
+  ) {
+    try {
+      await js.pull(stream, "me", expires);
+    } catch (err) {
+      assertEquals(err.code, code);
+    }
+  }
+
+  await expectError(0, ErrorCode.JetStream404NoMessages);
+  await expectError(1000, ErrorCode.JetStream408RequestTimeout);
+
+  await js.publish(subj);
+
+  // we expect a message
+  const a = await js.pull(stream, "me", 1000);
+  assertEquals(a.seq, 1);
+
+  await cleanup(ns, nc);
+});
+
+Deno.test("jetstream - pull error: max_waiting", async () => {
+  const { ns, nc } = await setup(jetstreamServerConf({}, true));
+  if (await notCompatible(ns, nc, "2.8.2")) {
+    return;
+  }
+
+  const { stream } = await initStream(nc);
+  const jsm = await nc.jetstreamManager();
+
+  await jsm.consumers.add(stream, {
+    durable_name: "me",
+    ack_policy: AckPolicy.Explicit,
+    max_waiting: 1,
+  });
+  const js = nc.jetstream();
+
+  async function expectError(
+    expires: number,
+    code: ErrorCode,
+  ): Promise<NatsError> {
+    const d = deferred<NatsError>();
+    try {
+      await js.pull(stream, "me", expires);
+    } catch (err) {
+      d.resolve(err);
+      assertEquals(err.code, code);
+    }
+    return d;
+  }
+  const errors = await Promise.all([
+    expectError(
+      3000,
+      ErrorCode.JetStream408RequestTimeout,
+    ),
+    // this a server bug, it should be a 408
```

Comment:
Why? You have the max number of pull requests waiting, so why is it a server bug to send a 409 to indicate that the max number of pulls is reached? Or are you saying that it should be a 408 Exceeded MaxWaiting instead of 409 Exceeded MaxWaiting?

Reply:
The server was canceling the first request (the one that was valid) and then hanging on the second. I removed the comment - I am OK with the current behaviour.

Reply:
The hanging was the bug with MaxWaiting==1 (which is now fixed). While fixing that issue, we also fixed the fact that a new request on max_waiting reached should fail instead of evicting a current one.